controller/src/dns/mod.rs

```diff
@@ -0,0 +1,218 @@
+mod error;
+
+use self::error::OrDnsError;
+use crate::plan::DnsPlan;
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use dashmap::DashMap;
+use dis_spawner::messages::dns::SetDnsRecord;
+use dis_spawner::types::ClusterName;
+use dis_spawner::Never;
+use dis_spawner::{messages::dns::DnsRecordType, nats::TypedNats};
+use error::Result;
+use std::net::Ipv4Addr;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::task::JoinHandle;
+use tokio::{
+    self,
+    net::{TcpListener, UdpSocket},
+};
+use trust_dns_server::client::rr::rdata::TXT;
+use trust_dns_server::{
+    authority::MessageResponseBuilder,
+    client::{
+        op::ResponseCode,
+        rr::{RData, Record, RecordType},
+    },
+    proto::op::Header,
+    server::{Request, RequestHandler, ResponseHandler, ResponseInfo},
+    ServerFuture,
+};
+
+struct ClusterDnsServer {
+    name_map: Arc<DashMap<(ClusterName, String, DnsRecordType), RData>>,
+    _handle: JoinHandle<anyhow::Result<()>>,
+}
+
+impl ClusterDnsServer {
+    pub async fn new(nc: &TypedNats) -> Self {
+        let nc = nc.clone();
+        let name_map: Arc<DashMap<(ClusterName, String, DnsRecordType), RData>> = Arc::default();
+
+        let handle = {
+            let name_map = name_map.clone();
+
+            tokio::spawn(async move {
+                let mut stream = nc
+                    .subscribe_jetstream(SetDnsRecord::subscribe_subject())
+                    .await?;
+
+                while let Some(v) = stream.next().await? {
+                    match v.kind {
+                        DnsRecordType::A => {
+                            let key = (v.cluster.clone(), v.name.clone(), DnsRecordType::A);
+                            let ip: Ipv4Addr = match v.value.parse() {
+                                Ok(v) => v,
+                                Err(error) => {
+                                    tracing::warn!(
+                                        ?error,
+                                        ip = v.value,
+                                        "Error parsing IP in SetDnsRecord request."
+                                    );
+                                    continue;
+                                }
+                            };
+                            let value = RData::A(ip);
+                            name_map.insert(key, value);
+                        }
+                        DnsRecordType::TXT => {
+                            let key = (v.cluster.clone(), v.name.clone(), DnsRecordType::TXT);
+                            let value = RData::TXT(TXT::new(vec![v.value]));
```

Comment:
if I understand correctly this sets the latest TXT record from the stream right? Would this not cause race conditions with multiple drones? It'd be good to write a test with multiple drones on the same cluster to check at what point this is an issue.